pennylane/qaoa/layers.py

```diff
@@ -0,0 +1,165 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Methods that define cost and mixer layers for use in QAOA workflows.
+"""
+import pennylane as qml
+from pennylane.operation import Tensor
+
+
+def _diagonal_terms(hamiltonian):
+    r"""Checks if all terms in a Hamiltonian are products of diagonal Pauli gates (PauliZ and Identity).
+
+    Args:
+        hamiltonian (.Hamiltonian): The Hamiltonian being checked
+
+    Returns:
+        bool: ``True`` if all terms are products of diagonal Pauli gates, ``False`` otherwise
+    """
+    val = True
+
+    for i in hamiltonian.ops:
+        i = Tensor(i) if isinstance(i.name, str) else i
+        for j in i.obs:
+            if j.name != "PauliZ" and j.name != "Identity":
+                val = False
+
+    return val
+
+
+def cost_layer(hamiltonian):
+    r"""Returns the QAOA cost layer corresponding to a cost Hamiltonian.
+
+    For the cost Hamiltonian :math:`H_C`, this is defined as the following unitary:
+
+    .. math:: U_C \ = \ e^{-i \gamma H_C}
+
+    where :math:`\gamma` is a variational parameter.
+
+    Args:
+        hamiltonian (.Hamiltonian): The cost Hamiltonian
+
+    Raises:
+        ValueError: if the terms of the supplied cost Hamiltonian are not exclusively products of diagonal Pauli gates
+
+    .. UsageDetails::
+
+        We define a cost Hamiltonian and pass it into ``cost_layer``:
+        and pass it into ``cost_layer``:
+
+        .. code-block:: python
```

Comment:
`python` refers to Python2, which is normally fine, except that Python 2 didn't have the `@` operator 😆 So syntax highlighting on the codeblock will fail
```suggestion
        .. code-block:: python3
```